src/Dictionaries/HashedDictionary.cpp

```diff
@@ -409,11 +431,103 @@ void HashedDictionary::updateData()
+        added_rows += map_ref->size();
+        map_ref->resize(added_rows);
+    }
+    else
+    {
+        const auto & map_ref = std::get<SparseCollectionPtrType<T>>(attribute.sparse_maps);
+        added_rows += map_ref->size();
+        map_ref->resize(added_rows);
+    }
+}
+void HashedDictionary::resize(size_t added_rows)
+{
+    if (!added_rows)
+        return;
+
+    for (auto & attribute : attributes)
+    {
+        switch (attribute.type)
+        {
+            case AttributeUnderlyingType::utUInt8:
+                resize<UInt8>(attribute, added_rows);
+                break;
+            case AttributeUnderlyingType::utUInt16:
+                resize<UInt16>(attribute, added_rows);
+                break;
+            case AttributeUnderlyingType::utUInt32:
+                resize<UInt32>(attribute, added_rows);
+                break;
+            case AttributeUnderlyingType::utUInt64:
+                resize<UInt64>(attribute, added_rows);
+                break;
+            case AttributeUnderlyingType::utUInt128:
+                resize<UInt128>(attribute, added_rows);
+                break;
+            case AttributeUnderlyingType::utInt8:
+                resize<Int8>(attribute, added_rows);
+                break;
+            case AttributeUnderlyingType::utInt16:
+                resize<Int16>(attribute, added_rows);
+                break;
+            case AttributeUnderlyingType::utInt32:
+                resize<Int32>(attribute, added_rows);
+                break;
+            case AttributeUnderlyingType::utInt64:
+                resize<Int64>(attribute, added_rows);
+                break;
+            case AttributeUnderlyingType::utFloat32:
+                resize<Float32>(attribute, added_rows);
+                break;
+            case AttributeUnderlyingType::utFloat64:
+                resize<Float64>(attribute, added_rows);
+                break;
+
+            case AttributeUnderlyingType::utDecimal32:
+                resize<Decimal32>(attribute, added_rows);
+                break;
+            case AttributeUnderlyingType::utDecimal64:
+                resize<Decimal64>(attribute, added_rows);
+                break;
+            case AttributeUnderlyingType::utDecimal128:
+                resize<Decimal128>(attribute, added_rows);
+                break;
+
+            case AttributeUnderlyingType::utString:
+                resize<StringRef>(attribute, added_rows);
+                break;
+        }
+    }
+}
+void HashedDictionary::updateSize()
+{
+    size_t current_new_size = new_size.exchange(0);
+    if (!current_new_size)
+        return;
+    resize(current_new_size);
+}
+
 void HashedDictionary::loadData()
 {
     if (!source_ptr->hasUpdateField())
     {
         auto stream = source_ptr->loadAll();
+
+        if (preallocate)
+        {
+            stream->setProgressCallback([this](const Progress & progress)
+            {
+                new_size += progress.total_rows_to_read;
```

Comment:
Now I think that it is dangerous to use `total_rows_to_read`.
If we use `clickhouse` source, it is possible to have a case with heavy filtration, e.g. when 90% of a table is filtered. Then we will definitively pre-allocate too much.

Reply:
Also, for other type of sources, I don't think we actually have good information about the result size.
I mean, it is unlikely that it will be better then sum for `block.rows()`, which we can calculate incrementally after reading `block` in loop below.

@azat could you change your code this way and check your test again?

Reply:
>Now I think that it is dangerous to use total_rows_to_read.
>If we use clickhouse source, it is possible to have a case with heavy filtration, e.g. when 90% of a table is filtered. Then we will definitively pre-allocate too much.

Indeed, and that's why it is optional

>@azat could you change your code this way and check your test again?

@KochetovNicolai  Do you have nay suggestions for `total_rows_to_read`?

Reply:
And resize by 65535 block does not give anything:

```
$ ./bench-sparse_hash-preallocate-block                                                    
sparse_hash/keys: 10000000
sparse_hash/insert: 5.47112
sparse_hash/find  : 1.5312
sparse_hash/maxrss: 174MiB 
sparse_hash/time:   7006.86 msec (user+sys)
$ ./bench-sparse_hash-preallocate                                               
sparse_hash/keys: 10000000
sparse_hash/insert: 3.71339
sparse_hash/find  : 1.53067
sparse_hash/maxrss: 174MiB 
sparse_hash/time:   5251.77 msec (user+sys)
$ ./bench-sparse_hash
sparse_hash/keys: 10000000
sparse_hash/insert: 5.48519
sparse_hash/find  : 1.49287
sparse_hash/maxrss: 174MiB 
sparse_hash/time:   6982.64 msec (user+sys)
```

Reply:
>If we use clickhouse source, it is possible to have a case with heavy filtration, e.g. when 90% of a table is filtered. Then we will definitively pre-allocate too much.

Or `IDictionarySource` can be extended to include something like `hasWhere` and then preallocate can be done only if there is no where, it will work, but even simple `WHERE` will break preallocation does not looks good to me

Reply:
@azat could you please show your test script?

Reply:
@KochetovNicolai if you are asking about benchmarks, then they are available [here](https://github.com/azat-archive/hashtable-bench)

README.md is a little bit outdated, so you need to compile it:

```
git clone --recursive https://github.com/azat-archive/hashtable-bench
cd hashtable-bench
mkdkr .cmake && cd .cmake
cmake -G Ninja -DCMAKE_BUILD_TYPE=Release ..
```

And you need only the following targets (not all included as submodules, i.e. folly, and it will fail to compile, hence it is better to compile only specific targets):

```
ninja -j1 bench-sparse_hash-run bench-sparse_hash-preallocate-run bench-sparse_hash-preallocate-block-run
```

(This will compile and run each benchmark, `-j1` to avoid bias, just in case)

Reply:
Thanks!

Reply:
But what I see in `PREALLOCATE_BLOCK` case in not quite what happens in `ClickHouse` has table. Reserve in out hash table rounds up by the power of 2. I don't think it is true for hash tables from test.

Also, I can't guarantee that `setProgressCallback` immediately has correct `total_rows_to_read`. So, we need to test your changes also. I will try to compare.